### PR TITLE
`TabPanel`: support manual tab activation

### LINKED ISSUE #45390: `TabPanel`: add support for manual activation of tabs
Currently the `TabPanel`'s tabpanels are activated automatically when the associated tab receives focus (referred to as "Tabs with automatic activation" on [the W3C website](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/)).

However, as discussed in https://github.com/WordPress/gutenberg/pull/44788#issuecomment-1293826203 and in https://github.com/WordPress/gutenberg/pull/45203#discussion_r1006643474, there instances in which the automatic activation doesn't fit the UX and keyboard navigation needs.

We should look into allowing the `TabPanel` to support manual activation as well, alongside automatic activation.

### PR DESCRIPTION
**Pull request by ciampo**
Closes #45390

## What?

Adds support for manual tab activation to the `TabPanel` component (see https://github.com/WordPress/gutenberg/issues/45390 and  [W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info)

## Why?

New feature for the `TabPanel` component, for enhanced accessibility when necessary.

## How?

- Introduced a new prop `hasManualTabActivation` (name TBD)
- The component keeps behaving as before when the new prop is not specified (backwards-compat)
- Added unit tests for both automatic and manual tab activation (including a necessary mock to test keyboard interactions)

## Testing Instructions

- Load the default `TabPanel` example in Storybook
- Click on the first tab, then use arrow key to switch tab. Note how the currently shown tabpanel changes as the arrow keys are pressed
- Via Storybook controls, change the value of the `hasManualTabActivation` prop to `true`
- Click on the first tab, then use arrow keys to navigate through the list of tabs. Notice how focusing a tab doesn't cause the currently shown tab panel to change.
- Make sure that the tabpanel changes only when the space or enter keys are pressed while focusing a tab.

## Screenshots or screencast 

https://user-images.githubusercontent.com/1083581/203551184-5f6f76d9-fa39-4128-bc72-dfd88d2b3f75.mp4

### CODE REVIEW

**Comment by ciampo** [comment]
@jasmussen and @jameskoster , this may require design feedback as I don't think that `TabPanel` styles ever accounted for a tab to be focused while not being active.

File: packages/components/src/tab-panel/README.md
Comment (ciampo) [inline packages/components/src/tab-panel/README.md]: Not entirely convinced about this new prop's name, open to suggestions!

File: packages/components/src/tab-panel/test/index.tsx
Comment (ntsekouras) [inline packages/components/src/tab-panel/test/index.tsx:46]: Curious why we need this.. I saw it's also the case for some other tests like `NavigableMenu`, but I'd love to understand better. Is it related with the testing library used?

File: packages/components/src/tab-panel/README.md
Comment (ntsekouras) [reply packages/components/src/tab-panel/README.md]: Maybe just `manualActivation`?

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR Marco! We just need to take a look at the non-active but focused styles, as you already mention. Other than that this looks good.

**Comment by jameskoster** [comment]
It's working for me:

https://user-images.githubusercontent.com/846565/203558051-9e94a613-e57b-4909-a5d5-0213c3702315.mp4

I don't have a great idea for the name either, but perhaps something that relates more to the states involved; `openPanelOnFocus`?

File: packages/components/src/tab-panel/test/index.tsx
Comment (ciampo) [reply packages/components/src/tab-panel/test/index.tsx:46]: `NavigableMenu` internally uses `focus.tabbable.find()` and `focus.focusable.find()` to search for the tabbable/focusable items — the logic requires a tabbable item [to be visible](https://github.com/WordPress/gutenberg/blob/feat/tab-panel-manual-tab-activation/packages/dom/build-module/focusable.js) — unfortunately, by default `js-dom` doesn't satisfy any of those checks, meaning that by default in our tests `NavigableMenu` won't move the selection to the next/previous tabs when interacting with the keyboard.

Mocking `getClientRects()` makes the above `isVisible` check pass and allows the test to work :)

File: packages/components/src/tab-panel/README.md
Comment (ciampo) [reply packages/components/src/tab-panel/README.md]: For boolean props I like to use names that start with `is` or `has` when possible (or at most, action-focused names), which makes it more obvious that the prop is a boolean — so maybe `hasManualActivation` or `activateTabsManually` ?

**Comment by ciampo** [comment]
> I don't have a great idea for the name either, but perhaps something that relates more to the states involved; `openPanelOnFocus`?

On a similar note, `ariakit` calls this feature `selectOnMove` — do we think it would be a better prop name?

**Comment by jameskoster** [comment]
> @jasmussen and @jameskoster , this may require design feedback as I don't think that TabPanel styles ever accounted for a tab to be focused while not being active.

We can probably consider this when we get to https://github.com/WordPress/gutenberg/issues/41904.

**Comment by chad1008** [comment]
One thing that's occurring to me is that `ariakit` and this change both default to displaying the tab on focus, which is good, but we come at it from opposite directions. To enable manual selection:
- `ariakit` wants `selectOnMove` to be `false`
- This PR will want `hasManualTabActivation` (or whatever we name it) to be `true`

So we'd eventually end up managing the `ariakit` prop with something like `selectOnMove = { ! hasManualTabActivation }`

That's doable, of course, but maybe it'd be more readable in the long run if we went (as @ciampo suggests) with the same `selectOnMove` prop name, and flipped our boolean accordingly. That way our component and `ariakit`'s implementation are completely in sync.

File: packages/components/src/tab-panel/test/index.tsx
Comment (ntsekouras) [reply packages/components/src/tab-panel/test/index.tsx:46]: So it's related to `js-dom`.. Thanks for elaborating Marco!

File: packages/components/src/tab-panel/README.md
Comment (ntsekouras) [reply packages/components/src/tab-panel/README.md]: Your suggestion for `selectOnMove` sound good to me.

**Review by alexstine — APPROVED** [review]
@ciampo Thanks for getting this done quickly. Hopefully, once this is merged, other PRs can progress and further enhancements can be made to code already in trunk.

Giving the PR an approval while the last few bits of feedback are addressed. I agree using the same prop name as we'll have in Ariakit could be useful.

File: packages/components/src/tab-panel/index.tsx
Comment (tyxla) [inline packages/components/src/tab-panel/index.tsx]: Nit: naming seems off here, did we mean to say `activateTabAutomatically`?

File: packages/components/src/tab-panel/test/index.tsx
Comment (tyxla) [inline packages/components/src/tab-panel/test/index.tsx]: Nit: should this be declared right before we override it in `beforeAll`? This would then be a `let` and we'd just set it in `beforeAll()`.

File: packages/components/src/tab-panel/test/index.tsx
Comment (tyxla) [inline packages/components/src/tab-panel/test/index.tsx]:
```suggestion
					hasManualTabActivation
```

**Review by tyxla — APPROVED** [review]
This is looking great, there's not much to add to what has been suggested in prior review already. 🚀

**Comment by ciampo** [comment]
I [renamed the prop](https://github.com/WordPress/gutenberg/pull/46004/commits/ebcb84d1c7f4a3e6d9dfbc8f568e02807bc2aa30) to `selectOnMove` as agreed, and addressed all remaining feedback.

I guess the last thing to decide is whether we're ok to merge this PR with the current way tabs look in manual activation mode, when they are focused but not active (see video in the description).

File: packages/components/src/tab-panel/index.tsx
Comment (ciampo) [reply packages/components/src/tab-panel/index.tsx]: Yup! Done in https://github.com/WordPress/gutenberg/pull/46004/commits/8ef6ea4fe0c64949ad0a4a30d48a0560ae37d498

File: packages/components/src/tab-panel/test/index.tsx
Comment (ciampo) [reply packages/components/src/tab-panel/test/index.tsx]: Done in https://github.com/WordPress/gutenberg/pull/46004/commits/3128c2684b703a8636e6af1c5a29aad01cde82eb

File: packages/components/src/tab-panel/test/index.tsx
Comment (ciampo) [reply packages/components/src/tab-panel/test/index.tsx]: No need for this anymore, since the new `selectOnMove` prop needs a `false` value here

File: packages/components/src/tab-panel/test/index.tsx
Comment (ciampo) [reply packages/components/src/tab-panel/test/index.tsx:46]: No worries! It took me a while to find out at first too 😅

File: packages/components/src/tab-panel/README.md
Comment (ciampo) [reply packages/components/src/tab-panel/README.md]: Done!

File: packages/components/src/tab-panel/README.md
Comment (tyxla) [inline packages/components/src/tab-panel/README.md]:
```suggestion
When `true`, the tab will be selected when receiving focus (automatic tab activation). When `false`, the tab will be selected only when clicked (manual tab activation). See the [official W3C docs](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/) for more info.
```

File: packages/components/src/tab-panel/types.ts
Comment (tyxla) [inline packages/components/src/tab-panel/types.ts]:
```suggestion
	 * When `true`, the tab will be selected when receiving focus (automatic tab
```

**Review by tyxla — APPROVED** [review]
Looks great 🚀

**Comment by ciampo** [comment]
> whether we're ok to merge this PR with the current way tabs look in manual activation mode

Given the absence of strong feedback against merging, I'll go ahead and merge this PR once CI is ✅ . As @jameskoster mentioned above:

> We can probably consider this when we get to #41904.

**Pull request merged**
